pkg/wireguard/agent/agent.go

```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
 
 	return wgtypes.NewKey(bytes)
 }
+
+// OnIPIdentityCacheChange implements ipcache.IPIdentityMappingListener
+func (a *Agent) OnIPIdentityCacheChange(modType ipcache.CacheModification, ipnet net.IPNet, oldHostIP, newHostIP net.IP,
+	oldID *identity.NumericIdentity, newID identity.NumericIdentity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata) {
+
+	// This function is invoked from the IPCache with the
+	// ipcache.IPIdentityCache lock held. We therefore need to be careful when
+	// calling into ipcache.IPIdentityCache from Agent to avoid potential
+	// deadlocks.
+	// We are only interested in IPCache entries where the hostIP is set, i.e.
+	// updates with oldHostIP set for deletion/updates and newHostIP set
+	// for newly created entries.
+	a.Lock()
+	defer a.Unlock()
+
+	// List of peers affected by this IPCache change
+	updatedPeers := []*peerConfig{}
+
+	// Remove IP from allowedIPs if it was deleted or the host IP changed
+	if modType == ipcache.Delete || !newHostIP.Equal(oldHostIP) {
```

Comment:
Shouldn't we handle the case when the host IP (aka node IP) changed only in `UpdatePeer()`? The latter will be invoked in any case upon the change.

If we could handle the change in `UpdatePeer()`, then both if's would become mutually exclusive, and we could get rid of `updatedPeers`.

Reply:
To be honest, I'm not sure what exactly happens when a node IP changes with regard to `UpdatePeer()`. I have to assume that a node IP change is not an atomic operation on the IPCache, i.e. we will get a bunch of `Upsert` events with `!newHostIP.Equal(oldHostIP)` here. I should probably test it (but I also understand that right now this is not yet fully implemented).

I think you actually convinced me that the current approach here is indeed not correct. Right now, we inherit the allowedIPs entries from the node before the IP change. Which means we suddenly have a different view than the IPCache, which might still have the old `hostIP` for some entries. In that case, we'll miss events such as `Delete` events because they still have the old host IP set, while we already assume the new one. :thinking:

Reply:
>  In that case, we'll miss events such as Delete events because they still have the old host IP set, while we already assume the new one.

Needs testing, but I *hope* that `UpdatePeer()` is called when the node IP addr changes. If it's guaranteed that `UpdatePeer()` happens before the IPCache updates, then we should easily handle this. But if there is no such guarantee, then :thinking:

Reply:
I have tested and verified that `UpdatePeer` is indeed invoked when the nodeIP changes. I have modified the code in a way that we always re-fetch the IPCache state whenever we detect a node IP change. So an IP change is basically identical to a node being added again. This way it does not matter if we see the `IPCache` events before or after we see the `UpdatePeer()` event.

However, I'm pretty sure that Cilium currently does not fully support node IP changes atm. The IPCache entries were not updated (which is out of scope for this PR).